Adds -h and -v flags for all examples

chaz-browne opened the pull request:
# Description

Improves examples by adding `-h` and a `-v` flags and usage statement to all examples.

First step in fixing #267

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

# How Has This Been Tested?
- [x] Comus (Broadwell),
- [x] Thompson (Haswell),
- [x] Alehouse (Power9)
- [x] Rhetoric (Skylake, single socket)
- [x] ARM Juno r2
- [x] Lassen with the CUDA host-config file build for NVML testing (different than alehouse).

# Checklist:

- [x] I have run `./scripts/check-code-format.sh` and confirm my code follows the style guidelines of variorum
- [x] I have added comments in my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass with my changes

chaz-browne committed:
Adds -h option and helper message for variorum-print-features-example.c

tpatki commented:
Hi @chaz-browne ! Welcome to the team! 

Thank you for making the examples better! Could you please do the following before we merge:

(1) Fix the formatting of the code files. 
You can run `scripts/check-code-format.sh` to fix the formatting issue in the failing test, it uses `astyle` to format the files. You may have to install `astyle`, but I believe it is available on comus.

(2) Could you test your build and examples on the power lab nodes and confirm their output (you can paste it here if you like)? I suggest at least 2-3 diverse architectures for testing, eg comus, thompson and rhetoric. 

Thank you!

chaz-browne committed:
Adds version parameter and fixes style issue.

chaz-browne committed:
Moves common #defines into variorum.h.

tpatki commented:
Hi @chaz-browne ! 

Usually, just as a good practice, one PR tracks one major/minor issue. I recommend you use this current PR for the examples issue, that is, the one adding `-h` to the examples.

You can open two separate PRs for the (1) `astyle` issue (although I need to check this, because it seems to work fine at my end), and (2) the moving `#defines` into variorum.h issue. This would mean you need to create separate branches at your end to keep making progress, and then open separate PRs, as these three issues are unrelated. 

That way, these can be reviewed separately and if needed by different people, and merged separately, depending on the change.

tpatki reviewed (changes requested):
Please create a separate PR for this. (Refer to my other comment)

chaz-browne committed:
Adds -h and -v to variorum-print-energy-example.c

rountree commented:
@tpatki @slabasan 
Ok, I think I know how I might do this.  I'm going to have @chaz-browne revert the bug fix and move it to its own branch.  That branch will correspond to the bug fix pr.

tpatki commented:
@rountree 
Yes. One way is to use `git reset HEAD~3` (assuming 3 previous commits), adding one commit at a time, pushing each of it out to a separate branch (you may need to reset again after each commit is in its separate branch). Another way might be to use `git cherry-pick` (https://www.atlassian.com/git/tutorials/cherry-pick) which I have less experience with. @slabasan may have better ideas.

Hopefully we don't have to do this again, sorry for the hassle.

tpatki commented:
@rountree @chaz-browne 
I created a separate PR for the `check-code-format` script, so that can be merged in first and soon. Thank you for identifying it! 
You can continue working on your examples and the headers -- please create two separate branches/PRs to track those so they can be reviewed independently.

slabasan commented:
Hi @chaz-browne, since we have merged in your fix for the check-code-format script, you can drop the fix from this branch and rebase against the dev branch. Also, please create a new PR for moving #defines into variorum.h issue, and we can get that merged into dev as well. Happy to help you with resolving your branches next week! Thank you for contributing to variorum!

rountree commented:
@slabasan @tpatki 
@chaz-browne and I will be working through this Monday morning.

chaz-browne committed:
Adds -h, -v options to example.

chaz-browne committed:
Adds -h, -v options to example.

chaz-browne committed:
Adds -h, -v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h, -v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h,-v options to example.

chaz-browne committed:
Adds -h,-v options in example.

tpatki commented:
@chaz-browne @rountree: Could you run `sh scripts/check-code-format.sh` and resubmit your PR? Looks like it fails the formatting test. Thank you!

chaz-browne committed:
Fixing astyle errors.

tpatki commented:
@rountree @slabasan 

Can we talk about the high-level goals of this PR? While I like the `-h` for the examples, I don't think including `variorum_config.h`  (which is a generated header I think) in all our examples to obtain the version of variorum is a good idea. We don't want our users including `variorum_config.h` imho -- our examples should be simple for most of variorum, where we only use `variorum.h`, otherwise it gets confusing.. 

If we really want to dump out version information -- it can be contained in the `variorum.h` API itself, and maybe the easier way to do is is to expose a `variorum_get_current_version()` API instead.

rountree commented:
@tpatki Thanks for the prompt replies, particularly when I know you have a lot of other stuff on your plate today (and thanks again to @slabasan for an hour of help on git this morning).

I think a `variorum_get_current_version()` is a reasonable solution.  Down the road we might want a dedicated function that prints the compile and runtime versions of hwloc and Jansson, but that's not a pressing need.

@chaz-browne Are you comfortable making that change?

tpatki commented:
Also, please test the PR on (just all relevant examples on each platform with the `-h` -- not all examples will be supported): 

- [ ] Comus (Broadwell), 
- [ ] Thompson (Haswell), 
- [ ] Alehouse (Power9)
- [ ] Rhetoric (Skylake, single socket)
- [ ] ARM Juno r2
- [ ] Lassen with the CUDA host-config file build.

These are the ones we have available. I wish we could also test actively on AMD and other Intel micro-architectures.

chaz-browne commented:
@tpatki Thank you for the review and suggestions! I'll be sure to work on these right now.

tpatki commented:
Hi @rountree @chaz-browne 
I wasn't asking for architecture specific functionality. I just want to make sure that the PR has been built and tested  thoroughly for expected behavior on all of the available architectures on the Power Lab before we merge the code in to mainstream. This is something @slabasan and I generally do to ensure we are not committing untested code (some of it will be, eg AMD) with all our PRs.

rountree commented:
@chaz-browne I caught up with @tpatki and have a much better idea of what she's looking for.  We'll talk tomorrow.

chaz-browne committed:
Implements suggestions by @tpatki

Includes "variorum_config.h" into "variorum.h"
Implemented "variorum_get_current_version();"
Adds "variorum_get_current_version" to single threaded examples.
-v,-h exit the example immediately.
Tested on: thompson, rhetoric, comus, sorrel.

chaz-browne committed:
Adds missing header file.

chaz-browne committed:
Made some fixes to usage prompt.

tpatki commented on src/examples/variorum-cap-best-effort-node-power-limit-example.c:
```diff
@@ -15,7 +15,7 @@ int main(int argc, char **argv)
     int node_pow_lim_watts = 500;
 
-    const char *usage = "%s [-hvl] -l power_lim_watts\n";
+    const char *usage = "Usage: %s [-hvl] -l power_lim_watts\n";
```
The options have listed `-l` twice: ` [-hvl] -l`. 

Similar typo in other places.

tpatki commented on src/examples/variorum-cap-best-effort-node-power-limit-example.c:
```diff
@@ -7,7 +7,6 @@
 #include <string.h>
-
 #include <variorum.h>
 
```
header file for getopt() is missing in many examples.

tpatki commented on src/examples/variorum-poll-power-to-stdout-example.c:
```diff
@@ -4,7 +4,7 @@
 
 #include <stdio.h>
-
+#include <variorum_config.h>
```
We don't need `variorum_config.h` here.

tpatki commented on src/examples/variorum-print-verbose-power-limit-example.c:
```diff
@@ -4,13 +4,34 @@
-
+#include <getopt.h>
+#include <stdlib.h>
+#include <variorum_config.h>
```
We don't need variorum_config.h here.

tpatki commented on src/variorum/variorum.h:
```diff
@@ -487,4 +487,8 @@ int variorum_get_node_power_domain_info_json(json_t *get_domain_obj);
 /// @brief Test for memory leaks.
 int variorum_tester(void);
 
+/// @brief To get a quoted variorum version, use QuoteMacro(VARIORUM_VERSION).
```
Doxygen picks up these comments to generate API docs.
Please add detailed documentation, eg return type of function (string with version number here).
Please add that it is supported on all architectures. 
The `QuoteMacro` part is internal now, so that should be removed from the comment here.

Also, add the new variorum function to API docs here: https://github.com/LLNL/variorum/blob/dev/src/docs/sphinx/api/print_functions.rst 

See examples of other variorum functions and API docs for what it would look like.

tpatki commented:
Thanks @chaz-browne and @rountree for all the hard work here! I left a few comments in the review, please take a look. 

- Build is failing as some examples don't include the `getopt.h` header file.
- In some examples, there's still a lingering reference to `variorum_config.h`, which we don't need anymore. 
A simple way to check for getopt as well as variorum_config is to `grep" in the examples folder to make sure all examples have been changed/updated. 
- In many of the examples that take an input (eg cap power limit or cap frequency), some options have been specified twice in the Usage string as such: 
`[-hvl] -l.`  <-- the second `-l` is not needed, correct? I also wonder if we should separate out optional and required arguments here, but we can leave that aside for now. 

- Please also add a more descriptive comment and add this to the API docs under `sphinx/api/print_functions.rst`, preferably toward the end of that file, that way doxygen picks it up and generates the docs for it.

chaz-browne committed:
Included <getopt.h> in headers.
Removed any lingering reference to <variorum_config.h>
Removed any repeating help options.

tpatki commented:
@chaz-browne @rountree 
Could you format (`sh scripts/check-code-format.sh`) and push the PR again? I think the other tests passes.

tpatki commented:
@rountree @chaz-browne 

In addition to formatting, you will also have to rebase and resolve conflicts for the two JSON examples, the API changed last week, so they are new examples and you'll have to edit your end to match it.

rountree-alt committed:
Adds -h and -v to topology example.

Changes -v to use variorum_get_current_version()

Formatting changes.

Improves -h option.

Updates command-line flags.

Updates parameter parsing.

Updates parameter processing.

Updates parameter parsing.

Fixup

rountree committed:
Adds docstrings to new get_version function.

slabasan committed:
cleanup examples

- add new line between standard and variorum header includes
- sort standard header includes alphabetically
- remove extra break statement after return
- use consistent flag ordering in usage statements
- remove last variorum_config.h header includes

tpatki commented on src/examples/variorum-print-verbose-power-limit-example.c:
```diff
@@ -4,13 +4,33 @@
 #include <stdio.h>
-
+#include <getopt.h>
+#include <stdlib.h>
```
Do we need <stdlib.h> here?

tpatki commented:
Also, from a previous review:

Please add this to the API docs under `sphinx/api/print_functions.rst`, preferably toward the end of that file, that way doxygen picks it up and generates the docs for it.

slabasan committed:
remove un-needed standard header includes

slabasan committed:
add variorum_get_current_version to Variorum print API docs

slabasan committed:
update usage statement, getopt parsing

slabasan committed:
update mpi and openmp examples

- update usage statement
- add -h and -v flags
- alphabetize standard headers
- remove unused headers

tpatki committed:
Rearrange order of print functions in API docs.
